bus-mapping/src/circuit_input_builder/execution.rs

```diff
@@ -459,3 +470,204 @@ impl Default for ExpEvent {
+}
+
+impl PrecompileEvents {
+    /// Get all ecrecover events.
+    pub fn get_ecrecover_events(&self) -> Vec<SignData> {
+        self.events
+            .iter()
+            .filter_map(|e| {
+                if let PrecompileEvent::Ecrecover(sign_data) = e {
+                    Some(sign_data)
+                } else {
+                    None
+                }
+            })
+            .cloned()
+            .collect()
+    }
+    /// Get all EcAdd events.
+    pub fn get_ec_add_events(&self) -> Vec<EcAddOp> {
+        self.events
+            .iter()
+            .filter_map(|e| {
+                if let PrecompileEvent::EcAdd(op) = e {
+                    Some(op)
+                } else {
+                    None
+                }
+            })
+            .cloned()
+            .collect()
+    }
+    /// Get all EcMul events.
+    pub fn get_ec_mul_events(&self) -> Vec<EcMulOp> {
+        self.events
+            .iter()
+            .filter_map(|e| {
+                if let PrecompileEvent::EcMul(op) = e {
+                    Some(op)
+                } else {
+                    None
+                }
+            })
+            .cloned()
+            .collect()
+    }
+    /// Get all EcPairing events.
+    pub fn get_ec_pairing_events(&self) -> Vec<EcPairingOp> {
+        self.events
+            .iter()
+            .cloned()
+            .filter_map(|e| {
+                if let PrecompileEvent::EcPairing(op) = e {
+                    Some(*op)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+}
+
+/// I/O from a precompiled contract call.
+#[derive(Clone, Debug)]
+pub enum PrecompileEvent {
+    /// Represents the I/O from Ecrecover call.
+    Ecrecover(SignData),
+    /// Represents the I/O from EcAdd call.
+    EcAdd(EcAddOp),
+    /// Represents the I/O from EcMul call.
+    EcMul(EcMulOp),
+    /// Represents the I/O from EcPairing call.
+    EcPairing(Box<EcPairingOp>),
+}
+
+impl Default for PrecompileEvent {
+    fn default() -> Self {
+        Self::Ecrecover(SignData::default())
+    }
+}
+
+/// EcAdd operation: P + Q = R
+#[derive(Clone, Debug)]
+pub struct EcAddOp {
+    /// First EC point.
+    pub p: G1Affine,
+    /// Second EC point.
+    pub q: G1Affine,
+    /// Addition of the first and second EC points.
+    pub r: G1Affine,
+}
+
+impl Default for EcAddOp {
+    fn default() -> Self {
+        let p = G1Affine::generator();
+        let q = G1Affine::generator();
+        let r = p.add(q).into();
+        Self { p, q, r }
+    }
+}
+
+impl EcAddOp {
+    /// Creates a new EcAdd op given the inputs and output.
+    pub fn new(p: G1Affine, q: G1Affine, r: G1Affine) -> Self {
+        assert_eq!(p.add(q), r.into());
+        Self { p, q, r }
+    }
+
+    /// Returns true if P == Q == Infinity.
+    pub fn is_zero(&self) -> bool {
+        self.p.is_identity().into() && self.q.is_identity().into()
```

Comment:
TODO(zhuo): sum(p, neg(p)) is also Infinity

Reply:
`Q == neg(P)` does panic here:
https://github.com/scroll-tech/halo2-lib/blob/develop/halo2-ecc/src/ecc/mod.rs#L714

Although it will be hard to handle this case in EVM Circuit, because we will need to prove `Q == neg(P)`.

So we are missing the handling of case where `Q == neg(P)`.

Reply:
I have handled this case in the commit:
https://github.com/scroll-tech/zkevm-circuits/pull/598/commits/e06308ff2e3f551806cbca0787137cd6c06b175d